pyspark_ai/tool.py

```diff
@@ -104,3 +104,104 @@ def _run(
+    Input to this tool is a str with a keyword of interest and a temp view name, output is the column name from the df
+    that contains the keyword.
+    Input should be pipe-separated, in the format: keyword|temp_view_name
+    """
+
+    def _run(
+        self, input: str, run_manager: Optional[CallbackManagerForToolRun] = None
+    ) -> str:
+        input_lst = input.split("|")
+
+        keyword = input_lst[0].lower()
+        temp_name = input_lst[1]
+
+        # get columns
+        df = self.spark.sql("select * from {}".format(temp_name))
+        col_lst = df.columns
+
+        for col in col_lst:
+            result = self.spark.sql(
+                "select * from {} where `{}` like '%{}%'".format(
+                    temp_name, col, keyword
+                )
+            )
+
+            if len(result.collect()) != 0:
+                return col
+
+        return ""
+
+    async def _arun(self, *args: Any, **kwargs: Any) -> Any:
+        raise NotImplementedError("ColumnQueryTool does not support async")
+
+
+class SimilarValueTool(BaseTool):
+    """Tool for running a Spark SQL query to rank the values in a column by similarity score to a keyword."""
+
+    spark: Union[SparkSession, ConnectSparkSession] = Field(exclude=True)
+    name = "similar_value"
+    description = """
+    This tool finds the semantically closest word to a keyword from a vector database, using the FAISS library.
+    Input to this tool is a pipe-separated string in this format: keyword|column_name|temp_view_name.
+    The temp_view_name will be queried in the column_name for the semantically closest word to the keyword.
+    """
+
+    vector_store: Any
+    stored_df_cols: set
+
+    def vector_similarity_search(self, search_text: str, col: str, temp_name: str):
+        from langchain.docstore.document import Document
+        from langchain.vectorstores import FAISS
+        from langchain.embeddings.openai import OpenAIEmbeddings
+
+        new_df = self.spark.sql("select distinct {} from {}".format(col, temp_name))
+        new_df_lst = [str(row[col]) for row in new_df.collect()]
+
+        # Create documents from col data
+        if (temp_name, col) not in self.stored_df_cols:
+            documents = []
+
+            for val in new_df_lst:
+                document = Document(page_content=str(val), metadata={"temp_name": temp_name, "col_name": col})
+                documents.append(document)
+
+            if not self.vector_store:
+                self.vector_store = FAISS.from_documents(documents, OpenAIEmbeddings())
+            else:
+                self.vector_store.add_documents(documents)
+
+            self.stored_df_cols.add((temp_name, col))
+
+        docs = self.vector_store.similarity_search(search_text)
+        return docs[0].page_content
+
+    def _run(
+        self, inputs: str, run_manager: Optional[CallbackManagerForToolRun] = None
+    ) -> str:
+        input_lst = inputs.split("|")
```

Comment:
Same as the comment in https://github.com/databrickslabs/pyspark-ai/pull/119/files#r1333630340